pkg/sdn/registry/netnamespace/migration.go

```
@@ -0,0 +1,105 @@
+			errList = append(errList, err)
+			continue
+		}
+	}
+
+	if len(errList) > 0 {
+		errList = append(errList, fmt.Errorf("Unexpected errors during NetNamespace migration, please retry the operation"))
+	}
+	return kerrors.NewAggregate(errList)
+}
+
+func (m *Migrate) deleteNetNamespace(netns *api.NetNamespace) {
+	var out api.NetNamespace
+	if err := m.etcdHelper.Delete(context.TODO(), path.Join(etcdNetNamespacePrefix, netns.ObjectMeta.Name), &out, nil); err != nil {
+		glog.Errorf("Failed to delete NetNamespace: %v, error: %v", netns, err)
+	}
+}
+
+func (m *Migrate) migrateNetNamespace(netns *api.NetNamespace) error {
+	// Get corresponding namespace for the NetNamespace object
+	ns, err := m.client.Get(netns.NetName)
+	// If corresponding namespace no longer exists, nothing to migrate in this case.
+	if errors.IsNotFound(err) {
+		m.deleteNetNamespace(netns)
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Get namespace %q failed: %v", netns.NetName, err)
+	}
+
+	// Set netid as namespace annotation
+	if err := netid.SetVNID(ns, netns.NetID); err != nil {
+		return fmt.Errorf("Set netid %d for namespace %q failed: %v", netns.NetID, ns.Name, err)
+	}
+
+	// Update namespace
+	updatedNs, err := m.client.Update(ns)
+	if errors.IsNotFound(err) {
+		m.deleteNetNamespace(netns)
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Update namespace %q failed: %v", ns.Name, err)
+	}
+
+	// Validate
+	if id, err := netid.GetVNID(updatedNs); err == nil {
+		if id == netns.NetID {
+			glog.Infof("Migrated netid %d from NetNamespace to annotation on namespace %q", netns.NetID, ns.Name)
+		} else {
+			return fmt.Errorf("Failed to migrate netid from NetNamespace to namespace %q, expected netid %d but got %d", ns.Name, netns.NetID, id)
+		}
+	} else {
+		return fmt.Errorf("Failed to migrate netid %d from NetNamespace to annotation on namespace %q", netns.NetID, ns.Name)
+	}
+
+	// Delete processed NetNamespace object
+	m.deleteNetNamespace(netns)
```

ignoring errors means we're not re-entrant, right?